hibernate-core/src/main/java/org/hibernate/query/sqm/internal/AbstractSqmSelectionQuery.java

```diff
@@ -73,8 +86,107 @@ protected void errorOrLogForPaginationWithCollectionFetch() {
 	}
 
 	public abstract SqmStatement<R> getSqmStatement();
+	protected abstract void setSqmStatement(SqmSelectStatement<R> statement);
 	public abstract DomainParameterXref getDomainParameterXref();
 	public abstract TupleMetadata getTupleMetadata();
+
+	private SqmSelectStatement<R> getSqmSelectStatement() {
+		final SqmStatement<R> sqmStatement = getSqmStatement();
+		if ( sqmStatement instanceof SqmSelectStatement ) {
+			return (SqmSelectStatement<R>) sqmStatement;
+		}
+		else {
+			throw new IllegalSelectQueryException( "Not a select query" );
+		}
+	}
+
+	@Override
+	public SelectionQuery<R> setOrder(List<Order<? super R>> orderList) {
+		SqmSelectStatement<R> sqm = getSqmSelectStatement();
+		sqm = sqm.copy( noParamCopyContext() );
+		final SqmSelectStatement<R> select = sqm;
+		sqm.orderBy( orderList.stream().map( order -> sortSpecification( select, order ) )
+				.collect( toList() ) );
+		// TODO: when the QueryInterpretationCache can handle caching criteria queries,
+		//       simply cache the new SQM as if it were a criteria query, and remove this:
+		getQueryOptions().setQueryPlanCachingEnabled( false );
+		setSqmStatement( sqm );
+		return this;
+	}
+
+
+	@Override
+	public SelectionQuery<R> setOrder(Order<? super R> order) {
+		SqmSelectStatement<R> sqm = getSqmSelectStatement();
+		sqm = sqm.copy( noParamCopyContext() );
+		sqm.orderBy( sortSpecification( sqm, order ) );
+		// TODO: when the QueryInterpretationCache can handle caching criteria queries,
+		//       simply cache the new SQM as if it were a criteria query, and remove this:
+		getQueryOptions().setQueryPlanCachingEnabled( false );
+		setSqmStatement( sqm );
+		return this;
+	}
+
+	@Override
+	public SelectionQuery<R> setPage(Page page) {
+		setMaxResults( page.getMaxResults() );
+		setFirstResult( page.getFirstResult() );
+		return this;
+	}
+
+	private SqmSelectStatement<KeyedResult<R>> paginateQuery(
+			List<Order<? super R>> keyDefinition, List<Comparable<?>> keyValues) {
+		@SuppressWarnings("unchecked")
+		final SqmSelectStatement<KeyedResult<R>> sqm =
+				(SqmSelectStatement<KeyedResult<R>>)
+						getSqmSelectStatement().copy( noParamCopyContext() );
+		final NodeBuilder builder = sqm.nodeBuilder();
+		//TODO: find a better way handle parameters
+		builder.setCriteriaValueHandlingMode(ValueHandlingMode.INLINE);
```

Comment:
Why do you want to inline values? The default is to use JDBC parameters which should be fine.

Reply:
If this is for testing, you can configure that in your test by using

```java
@ServiceRegistry(settings = {
		@Setting(name = AvailableSettings.CRITERIA_VALUE_HANDLING_MODE, value = "inline")
})
```

or

```java
@Jpa(
		properties = @Setting(name = AvailableSettings.CRITERIA_VALUE_HANDLING_MODE, value = "inline")
)
```